Pan on zoom

jawadst opened the pull request:
See https://github.com/swimlane/ngx-graph/issues/52

isaacplmann commented:
This is a great feature, and I'd absolutely merge it, but it doesn't quite work.

If you zoom in, move your cursor and then zoom in or out, the zoom is no longer centered on your cursor.  You probably need to do some processing on the mouse position to get into the same coordinate plane as the graph.  Also, test dragging the graph around before zooming in and out.

jawadst commented:
@isaacplmann 
Thanks for the review!

I have pushed a new version that fixes the issue you mentioned and the zoom should now work correctly. There was indeed an issue with the mouse position that had to be converted to the SVG position and my algorithm for panzoom was off (basically you have to pan, zoom then pan back).

Also, I switched to using a transformation matrix which makes the computation easier to model but does involve a lot of changes to the code. I have kept backward compatibility with panOffset/zoomLevel but they are not used internally anymore and everything relies on the matrix (which is what I have seen other libraries do). I have added zoom/zoomTo/pan/panTo functions to make interacting with the component programmatically easier.

Let me know what you think.

isaacplmann commented:
Excellent!  This looks great.
